crates/core_simd/examples/nbody.rs

```diff
@@ -0,0 +1,186 @@
+            0.,
+        ]),
+        v: f64x4::from_array([
+            -2.76742510726862411e-03 * DAYS_PER_YEAR,
+            4.99852801234917238e-03 * DAYS_PER_YEAR,
+            2.30417297573763929e-05 * DAYS_PER_YEAR,
+            0.,
+        ]),
+        mass: 2.85885980666130812e-04 * SOLAR_MASS,
+    },
+    // uranus:
+    Body {
+        x: f64x4::from_array([
+            1.28943695621391310e+01,
+            -1.51111514016986312e+01,
+            -2.23307578892655734e-01,
+            0.,
+        ]),
+        v: f64x4::from_array([
+            2.96460137564761618e-03 * DAYS_PER_YEAR,
+            2.37847173959480950e-03 * DAYS_PER_YEAR,
+            -2.96589568540237556e-05 * DAYS_PER_YEAR,
+            0.,
+        ]),
+        mass: 4.36624404335156298e-05 * SOLAR_MASS,
+    },
+    // neptune:
+    Body {
+        x: f64x4::from_array([
+            1.53796971148509165e+01,
+            -2.59193146099879641e+01,
+            1.79258772950371181e-01,
+            0.,
+        ]),
+        v: f64x4::from_array([
+            2.68067772490389322e-03 * DAYS_PER_YEAR,
+            1.62824170038242295e-03 * DAYS_PER_YEAR,
+            -9.51592254519715870e-05 * DAYS_PER_YEAR,
+            0.,
+        ]),
+        mass: 5.15138902046611451e-05 * SOLAR_MASS,
+    },
+];
+
+pub fn offset_momentum(bodies: &mut [Body; N_BODIES]) {
+    let (sun, rest) = bodies.split_at_mut(1);
+    let sun = &mut sun[0];
+    for body in rest {
+        let m_ratio = body.mass / SOLAR_MASS;
+        sun.v -= body.v * m_ratio;
+    }
+}
+
+pub fn energy(bodies: &[Body; N_BODIES]) -> f64 {
+    let mut e = 0.;
+    for i in 0..N_BODIES {
+        let bi = &bodies[i];
+        e += bi.mass * (bi.v * bi.v).horizontal_sum() * 0.5;
+        for bj in bodies.iter().take(N_BODIES).skip(i + 1) {
+            let dx = bi.x - bj.x;
+            e -= bi.mass * bj.mass / (dx * dx).horizontal_sum().sqrt()
+        }
+    }
+    e
+}
+
+pub fn advance(bodies: &mut [Body; N_BODIES], dt: f64) {
+    const N: usize = N_BODIES * (N_BODIES - 1) / 2;
+
+    // compute distance between bodies:
+    let mut r = [f64x4::splat(0.); N];
+    {
+        let mut i = 0;
+        for j in 0..N_BODIES {
+            for k in j + 1..N_BODIES {
+                r[i] = bodies[j].x - bodies[k].x;
+                i += 1;
+            }
+        }
+    }
+
+    let mut mag = [0.0; N];
+    let mut i = 0;
+    while i < N {
+        let d2s = f64x2::from_array([
+            (r[i] * r[i]).horizontal_sum(),
+            (r[i + 1] * r[i + 1]).horizontal_sum(),
+        ]);
+        let dmags = f64x2::splat(dt) / (d2s * d2s.sqrt());
+        mag[i] = dmags[0];
+        mag[i + 1] = dmags[1];
+        i += 2;
+    }
+
+    i = 0;
+    for j in 0..N_BODIES {
+        for k in j + 1..N_BODIES {
+            let f = r[i] * mag[i];
+            bodies[j].v -= f * bodies[k].mass;
+            bodies[k].v += f * bodies[j].mass;
+            i += 1
+        }
+    }
```

Comment:
Ideally, for cases like this, we would declare the value of `i` in a formulaic fashion here instead of mutating it, but using `let mut i = 0;` here would be fine, to confine the mutation to this passage of code instead of the... very "C" idiom of reusing variables from earlier.

Reply:
Oh, that is a nice way of putting it. I will try to enclose those mutations to smaller scopes. Good catch.